File: java/src/main/java/io/github/muthuishere/toolnexus/Toolkit.java
package io.github.muthuishere.toolnexus;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Toolkit — aggregates dynamic MCP tools + the skill tool + custom/native/http
 * tools into a single uniform list, with provider adapters and a single
 * {@code execute()} router. Mirrors the JS reference ({@code js/src/toolkit.ts}).
 */
public final class Toolkit implements AutoCloseable {
    private final McpSource mcp;       // may be null
    private final SkillSource skill;   // may be null
    /** A2A profile read from a top-level {@code a2a} config block; {@link #serve} falls back to this. */
    private final A2AServer.A2AConfig a2aConfig; // may be null
    /** MCP serve profile read from a top-level {@code mcpServer} (singular) config block. */
    private final McpServe.MCPServeConfig mcpServerConfig; // may be null
    private final Map<String, Tool> byName = new LinkedHashMap<>();

    public static final class Options {
        public Object mcpConfig;            // path string, raw JSON string, or parsed Map
        public List<String> skillsDir;      // one or more skill roots
        /** Skills supplied as data, bypassing the filesystem (§3, S1). */
        public List<SkillSource.SkillDef> skills;
        /** Lazy provider of data skills, resolved once at build; failure isolated (§3, S1). */
        public java.util.function.Supplier<List<SkillSource.SkillDef>> skillProvider;
        /** Per-agent skill allowlist keyed on name; same semantics as the MCP tools filter (§3, S2). */
        public Map<String, Boolean> skillsFilter;
        /** Sibling-file sample cap: 0 ⇒ default 10, n>0 ⇒ cap, -1 ⇒ omit &lt;skill_files&gt; (§3, S5). */
        public int skillSampleLimit;
        public List<Tool> extraTools;       // custom/native/http tools
        public List<Object> annotatedObjects; // objects scanned via Tools.fromObject
        /** Built-in tools (§4A). On by default. false | {disabled:true} | {enabled:false} => off. */
        public Object builtins;
        /** Remote A2A agents — each advertised skill becomes a tool (source:"a2a"). */
        public List<A2A.Agent> agents;
        /**
         * Host resolver for out-of-band input (§10). When set, connected MCP servers may elicit
         * input from the human mid-{@code tools/call} and it is bridged onto this {@code waitFor}
         * (form→kind:"input", URL→kind:"authorization"). Typically the same function passed to the
         * client. Omit ⇒ MCP elicitation is not advertised.
         */
        public java.util.function.Function<Request, Answer> waitFor;
        /**
         * Drop tools by their FINAL exposed name across every source (MCP {@code server_tool},
         * builtins, native, A2A) — the simple "just turn these off" list, applied after aggregation.
         * Composes with the per-server/per-skill filters. Empty/null ⇒ nothing disabled.
         */
        public List<String> disableTools;
        /**
         * Drop skills by name from the {@code skill} catalog — sugar over a drop-list in
         * {@link #skillsFilter} (each name folded in as {@code false}). Empty/null ⇒ nothing disabled.
         */
        public List<String> disableSkills;
        /**
         * §2 Gap 3: a cooperative cancellation signal threaded into the MCP load (connect/init/list).
         * Null ⇒ today's behavior. See {@link McpSource.CancelSignal}.
         */
        public McpSource.CancelSignal cancelSignal;

        public Options mcpConfig(Object v) { this.mcpConfig = v; return this; }
        public Options skillsDir(String... v) { this.skillsDir = List.of(v); return this; }
        public Options skillsDir(List<String> v) { this.skillsDir = v; return this; }
        public Options skills(List<SkillSource.SkillDef> v) { this.skills = v; return this; }
        public Options skillProvider(java.util.function.Supplier<List<SkillSource.SkillDef>> v) { this.skillProvider = v; return this; }
        public Options skillsFilter(Map<String, Boolean> v) { this.skillsFilter = v; return this; }
        public Options skillSampleLimit(int v) { this.skillSampleLimit = v; return this; }
        public Options extraTools(List<Tool> v) { this.extraTools = v; return this; }
        public Options extraTools(Tool... v) { this.extraTools = List.of(v); return this; }
        public Options annotatedObjects(Object... v) { this.annotatedObjects = List.of(v); return this; }
        public Options builtins(Object v) { this.builtins = v; return this; }
        public Options agents(List<A2A.Agent> v) { this.agents = v; return this; }
        public Options agents(A2A.Agent... v) { this.agents = List.of(v); return this; }
        public Options waitFor(java.util.function.Function<Request, Answer> v) { this.waitFor = v; return this; }
        public Options disableTools(List<String> v) { this.disableTools = v; return this; }
        public Options disableTools(String... v) { this.disableTools = List.of(v); return this; }
        public Options disableSkills(List<String> v) { this.disableSkills = v; return this; }
        public Options disableSkills(String... v) { this.disableSkills = List.of(v); return this; }
        public Options cancelSignal(McpSource.CancelSignal v) { this.cancelSignal = v; return this; }
    }

    private Toolkit(McpSource mcp, SkillSource skill, List<Tool> builtins, List<Tool> agents, List<Tool> extraTools,
                   A2AServer.A2AConfig a2aConfig, McpServe.MCPServeConfig mcpServerConfig, List<String> disableTools) {
        this.mcp = mcp;
        this.skill = skill;
        this.a2aConfig = a2aConfig;
        this.mcpServerConfig = mcpServerConfig;
        // DisableTools drops tools by their final exposed name across every source.
        Set<String> disabled = new HashSet<>();
        if (disableTools != null) disabled.addAll(disableTools);
        // Builtins are the lowest-precedence source: a host extraTools entry with
        // the same name shadows a builtin (SPEC §4). Drop shadowed builtins up
        // front, then apply the normal first-wins dedupe for MCP/skill/agents/extras.
        Set<String> extraNames = new HashSet<>();
        for (Tool t : extraTools) extraNames.add(t.name());
        List<Tool> all = new ArrayList<>();
        if (mcp != null) all.addAll(mcp.tools());
        if (skill != null) all.add(skill.tool());
        for (Tool b : builtins) {
            if (!extraNames.contains(b.name())) all.add(b);
        }
        all.addAll(agents);
        all.addAll(extraTools);
        for (Tool t : all) {
            if (disabled.contains(t.name())) continue;
            if (byName.containsKey(t.name())) {
                System.err.println("[toolnexus] duplicate tool name \"" + t.name() + "\" — keeping first");
                continue;
            }
            byName.put(t.name(), t);
        }
    }

    public static Toolkit create(Options opts) {
        // §2 Gap 3: thread the caller's cancellation signal into the MCP connect/init/list.
        McpSource mcp = opts.mcpConfig != null
                ? McpSource.load(opts.mcpConfig, opts.waitFor, opts.cancelSignal) : null;
        // Skills come from directories, in-memory data, and/or a lazy provider (§3,
        // S1). The provider is resolved here and merged with the data list; a
        // provider failure is isolated so other sources still load.
        SkillSource skill = null;
        if (opts.skillsDir != null || opts.skills != null || opts.skillProvider != null) {
            java.util.List<SkillSource.SkillDef> dataDefs = new java.util.ArrayList<>();
            if (opts.skills != null) dataDefs.addAll(opts.skills);
            if (opts.skillProvider != null) {
                try {
                    java.util.List<SkillSource.SkillDef> provided = opts.skillProvider.get();
                    if (provided != null) dataDefs.addAll(provided);
                } catch (RuntimeException e) {
                    System.err.println("[toolnexus] skill provider failed: " + e.getMessage());
                }
            }
            // DisableSkills is sugar over a drop-list: fold each name into the filter as false
            // (without overriding an explicit skillsFilter entry).
            Map<String, Boolean> filter = opts.skillsFilter;
            if (opts.disableSkills != null && !opts.disableSkills.isEmpty()) {
                Map<String, Boolean> merged = new LinkedHashMap<>();
                if (opts.skillsFilter != null) merged.putAll(opts.skillsFilter);
                for (String n : opts.disableSkills) merged.putIfAbsent(n, false);
                filter = merged;
            }
            SkillSource.LoadOptions so = new SkillSource.LoadOptions();
            so.dirs = opts.skillsDir;
            so.skills = dataDefs.isEmpty() ? null : dataDefs;
            so.filter = filter;
            so.sampleLimit = opts.skillSampleLimit;
            skill = SkillSource.loadWith(so);
        }

        List<Tool> extras = new ArrayList<>();
        if (opts.extraTools != null) extras.addAll(opts.extraTools);
        if (opts.annotatedObjects != null) {
            for (Object o : opts.annotatedObjects) {
                extras.addAll(Tools.fromObject(o));
            }
        }

        // The toggle comes from the `builtins` option, or a top-level `builtins`
        // key on a parsed config Map — same precedence as MCP's isEnabled.
        Object builtinsCfg = opts.builtins;
        if (builtinsCfg == null && opts.mcpConfig instanceof Map
                && ((Map<?, ?>) opts.mcpConfig).containsKey("builtins")) {
            builtinsCfg = ((Map<?, ?>) opts.mcpConfig).get("builtins");
        }
        List<Tool> builtins = BuiltinTools.select(builtinsCfg);

        // Remote A2A agents come from the `agents:[...]` option plus a top-level
        // `agents` block on a parsed config Map (mirrors mcpServers). Each is
        // resolved to its skill tools; a failing agent never breaks the toolkit.
        List<A2A.Agent> descriptors = new ArrayList<>();
        if (opts.agents != null) descriptors.addAll(opts.agents);
        if (opts.mcpConfig instanceof Map && ((Map<?, ?>) opts.mcpConfig).containsKey("agents")) {
            Object block = ((Map<?, ?>) opts.mcpConfig).get("agents");
            @SuppressWarnings("unchecked")
            Map<String, Object> agentsBlock = block instanceof Map ? (Map<String, Object>) block : null;
            descriptors.addAll(A2A.parseAgentsConfig(agentsBlock));
        }
        List<Tool> agentTools = new ArrayList<>();
        for (A2A.Agent ag : descriptors) {
            try {
                agentTools.addAll(A2A.agentTools(ag));
            } catch (Exception e) {
                System.err.println("[toolnexus] A2A agent \"" + ag.card() + "\" failed: "
                        + (e.getMessage() == null ? String.valueOf(e) : e.getMessage()));
            }
        }

        // A2A inbound profile: a top-level `a2a` block on a parsed config Map
        // (mirrors builtins/agents). serve() prefers its inline `a2a` over this.
        A2AServer.A2AConfig a2aConfig = null;
        if (opts.mcpConfig instanceof Map && ((Map<?, ?>) opts.mcpConfig).containsKey("a2a")) {
            Object block = ((Map<?, ?>) opts.mcpConfig).get("a2a");
            if (block instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> a2aBlock = (Map<String, Object>) block;
                a2aConfig = A2AServer.A2AConfig.fromMap(a2aBlock);
            }
        }

        // MCP inbound profile: a top-level `mcpServer` (singular) block on a parsed
        // config Map — distinct from the client-side `mcpServers` block. serve()
        // prefers its inline `mcp` over this.
        McpServe.MCPServeConfig mcpServerConfig = null;
        if (opts.mcpConfig instanceof Map && ((Map<?, ?>) opts.mcpConfig).containsKey("mcpServer")) {
            Object block = ((Map<?, ?>) opts.mcpConfig).get("mcpServer");
            if (block instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> mcpBlock = (Map<String, Object>) block;
                mcpServerConfig = McpServe.MCPServeConfig.fromMap(mcpBlock);
            }
        }

        return new Toolkit(mcp, skill, builtins, agentTools, extras, a2aConfig, mcpServerConfig, opts.disableTools);
    }

    public List<Tool> tools() {
        return new ArrayList<>(byName.values());
    }

    public Tool get(String name) {
        return byName.get(name);
    }

    /** Add native/http/custom tools at runtime. First-name-wins; warn on dup. Chainable. */
    public Toolkit register(Tool... tools) {
        for (Tool t : tools) {
            if (byName.containsKey(t.name())) {
                System.err.println("[toolnexus] duplicate tool name \"" + t.name() + "\" — keeping first");
                continue;
            }
            byName.put(t.name(), t);
        }
        return this;
    }

    /**
     * Fetch a remote A2A agent's card and register its skills as tools. First-name-wins
     * dedupe. Mirrors the JS reference {@code toolkit.addAgent}.
     */
    public Toolkit addAgent(A2A.Agent ag) {
        try {
            return register(A2A.agentTools(ag).toArray(new Tool[0]));
        } catch (Exception e) {
            throw new RuntimeException("A2A addAgent failed: "
                    + (e.getMessage() == null ? String.valueOf(e) : e.getMessage()), e);
        }
    }

    /** Convenience: add an A2A agent from a bare Agent Card URL. */
    public Toolkit addAgent(String cardUrl) {
        return addAgent(A2A.agent(cardUrl));
    }

    /** Options for {@link #serve} — the client loop plus the opt-in A2A/MCP profiles + callbacks. */
    public static final class ServeOptions {
        public LlmClient client;
        public A2AServer.A2AConfig a2a;      // null ⇒ fall back to the toolkit's config block
        public A2AServer.OnTask onTask;
        public McpServe.MCPServeConfig mcp;  // null ⇒ fall back to the toolkit's `mcpServer` block
        public McpServe.OnCall onCall;

        public ServeOptions client(LlmClient v) { this.client = v; return this; }
        public ServeOptions a2a(A2AServer.A2AConfig v) { this.a2a = v; return this; }
        public ServeOptions onTask(A2AServer.OnTask v) { this.onTask = v; return this; }
        public ServeOptions mcp(McpServe.MCPServeConfig v) { this.mcp = v; return this; }
        public ServeOptions onCall(McpServe.OnCall v) { this.onCall = v; return this; }
    }

    /**
     * Serve this toolkit as an agent over HTTP. When the {@code a2a} profile is present
     * (inline, or a top-level {@code a2a} config block the toolkit was built from), it
     * mounts the A2A Agent Card ({@code /.well-known/agent-card.json}, built from skills)
     * and a JSON-RPC endpoint ({@code SendMessage} + {@code GetTask}) fulfilled by the client.
     * A message's A2A {@code contextId} keys the conversation via {@code client.ask}, so a
     * peer's turns are remembered through the client's {@link LlmClient.ConversationStore}.
     * When {@code a2a} is absent, no A2A routes are mounted. Returns a stoppable handle.
     * Mirrors the JS reference {@code toolkit.serve}.
     */
    public A2AServer.ServeHandle serve(String addr, ServeOptions opts) {
        A2AServer.A2AConfig a2a = opts.a2a != null ? opts.a2a : this.a2aConfig;
        McpServe.MCPServeConfig mcp = opts.mcp != null ? opts.mcp : this.mcpServerConfig;
        List<SkillSource.SkillInfo> skills = skill != null
                ? new ArrayList<>(skill.skills().values())
                : new ArrayList<>();
        LlmClient client = opts.client;
        try {
            // A message's A2A contextId keys the conversation via client.ask, so a peer's turns
            // are remembered through the client's ConversationStore; no contextId ⇒ stateless run.
            return A2AServer.start(addr, a2a, skills,
                    (text, contextId) -> (contextId != null && !contextId.isEmpty())
                            ? client.ask(text, this, contextId)
                            : client.run(text, this),
                    opts.onTask,
                    mcp, mcp != null ? tools() : null, opts.onCall);
        } catch (java.io.IOException e) {
            throw new RuntimeException("serve failed: " + e.getMessage(), e);
        }
    }

    public ToolResult execute(String name, Map<String, Object> args, ToolContext ctx) {
        Tool tool = byName.get(name);
        if (tool == null) {
            return ToolResult.error("Unknown tool: " + name);
        }
        return tool.execute(args == null ? Map.of() : args, ctx);
    }

    public ToolResult execute(String name, Map<String, Object> args) {
        return execute(name, args, null);
    }

    /** Markdown skill catalog for the system prompt. */
    public String skillsPrompt() {
        return skill != null ? skill.prompt() : "";
    }

    /** server -> "connected" | "failed" | "disabled". */
    public Map<String, String> mcpStatus() {
        return mcp != null ? mcp.status() : Map.of();
    }

    public List<Map<String, Object>> toOpenAI() {
        return Adapters.toOpenAI(tools());
    }

    public List<Map<String, Object>> toAnthropic() {
        return Adapters.toAnthropic(tools());
    }

    public List<Map<String, Object>> toGemini() {
        return Adapters.toGemini(tools());
    }

    @Override
    public void close() {
        if (mcp != null) mcp.close();
    }
}

File: csharp/src/Toolnexus/Toolkit.cs
namespace Toolnexus;

/// <summary>
/// Toolkit — aggregates dynamic MCP tools + the skill tool + custom/native/http
/// tools into a single uniform list, with provider adapters and a single
/// <see cref="ExecuteAsync"/> router.
/// </summary>
public sealed class Toolkit : IAsyncDisposable
{
    private readonly McpSource? _mcp;
    private readonly SkillSource? _skill;
    private readonly Dictionary<string, ITool> _byName = new();

    /// <summary>A2A profile read from a top-level <c>a2a</c> config block (ServeAsync falls back to this).</summary>
    private readonly A2AConfig? _a2aConfig;

    /// <summary>MCP serve profile read from a top-level <c>mcpServer</c> config block (ServeAsync falls back to this).</summary>
    private readonly MCPServeConfig? _mcpServerConfig;

    public sealed class Options
    {
        /// <summary>Path string, raw JSON string, or parsed config dictionary.</summary>
        public object? McpConfig { get; set; }

        /// <summary>One or more skill roots.</summary>
        public List<string>? SkillsDir { get; set; }

        /// <summary>Custom/native/http tools.</summary>
        public List<ITool>? ExtraTools { get; set; }

        /// <summary>Objects scanned via <see cref="Tools.FromObject"/>.</summary>
        public List<object>? AnnotatedObjects { get; set; }

        /// <summary>
        /// Built-in tools (§4A). On by default. <c>false</c> | <c>{disabled:true}</c> |
        /// <c>{enabled:false}</c> ⇒ off; a <c>{tools:{name:bool,...}}</c> map disables named
        /// tools on the all-on baseline. Accepts a <see cref="bool"/> or an
        /// <c>IDictionary&lt;string, object?&gt;</c>. <c>null</c> = not set (defaults on, or falls
        /// back to a top-level <c>builtins</c> key on a parsed <see cref="McpConfig"/> object).
        /// </summary>
        public object? Builtins { get; set; }

        /// <summary>Remote A2A agents — each advertised skill becomes a tool (source <c>"a2a"</c>).</summary>
        public List<Agent>? Agents { get; set; }

        /// <summary>
        /// Host resolver for out-of-band input (§10). When set, connected MCP servers may elicit
        /// input from the human mid-<c>tools/call</c> and it is bridged onto this <c>WaitFor</c>
        /// (form→kind:"input", URL→kind:"authorization"). Typically the same function passed to the
        /// client. Omit ⇒ MCP elicitation is not advertised.
        /// </summary>
        public Func<Request, Task<Answer>>? WaitFor { get; set; }

        public Options WithMcpConfig(object? v) { McpConfig = v; return this; }
        public Options WithSkillsDir(params string[] v) { SkillsDir = v.ToList(); return this; }
        public Options WithExtraTools(params ITool[] v) { ExtraTools = v.ToList(); return this; }
        public Options WithAnnotatedObjects(params object[] v) { AnnotatedObjects = v.ToList(); return this; }
        public Options WithBuiltins(object? v) { Builtins = v; return this; }
        public Options WithAgents(params Agent[] v) { Agents = v.ToList(); return this; }
        public Options WithWaitFor(Func<Request, Task<Answer>> v) { WaitFor = v; return this; }
    }

    private Toolkit(McpSource? mcp, SkillSource? skill, List<ITool> builtins, List<ITool> agents, List<ITool> extraTools, A2AConfig? a2aConfig, MCPServeConfig? mcpServerConfig)
    {
        _mcp = mcp;
        _skill = skill;
        _a2aConfig = a2aConfig;
        _mcpServerConfig = mcpServerConfig;
        // Builtins are the lowest-precedence source: a host extraTools entry with the same name
        // shadows a builtin (SPEC §4). Drop shadowed builtins up front, then apply the normal
        // first-wins dedupe for MCP/skill/agents/extras. Order: MCP → skill → builtin → agents → extraTools.
        var extraNames = new HashSet<string>(extraTools.Select(t => t.Name));
        var activeBuiltins = builtins.Where(b => !extraNames.Contains(b.Name)).ToList();
        var all = new List<ITool>();
        if (mcp != null) all.AddRange(mcp.Tools);
        if (skill != null) all.Add(skill.Tool);
        all.AddRange(activeBuiltins);
        all.AddRange(agents);
        all.AddRange(extraTools);
        foreach (var t in all) Add(t);
    }

    public static async Task<Toolkit> CreateAsync(Options opts)
    {
        var mcp = opts.McpConfig != null ? await McpSource.LoadAsync(opts.McpConfig, opts.WaitFor).ConfigureAwait(false) : null;
        var skill = opts.SkillsDir != null ? SkillSource.Load(opts.SkillsDir) : null;

        var extras = new List<ITool>();
        if (opts.ExtraTools != null) extras.AddRange(opts.ExtraTools);
        if (opts.AnnotatedObjects != null)
            foreach (var o in opts.AnnotatedObjects)
                extras.AddRange(global::Toolnexus.Tools.FromObject(o));

        // The toggle comes from the `Builtins` option, or a top-level `builtins` key on a parsed
        // config object (same precedence as MCP's isEnabled).
        var builtinsCfg = opts.Builtins;
        if (builtinsCfg == null && opts.McpConfig is IDictionary<string, object?> mc && mc.ContainsKey("builtins"))
            builtinsCfg = mc.Get("builtins");
        var builtins = BuiltinTools.Select(builtinsCfg);

        // Remote A2A agents come from the `Agents` option plus a top-level `agents` block on a
        // parsed config object (mirrors mcpServers). Each is resolved to its skill tools; a
        // failing agent never breaks the toolkit (isolated + logged, like MCP).
        var agentDescriptors = new List<Agent>();
        if (opts.Agents != null) agentDescriptors.AddRange(opts.Agents);
        if (opts.McpConfig is IDictionary<string, object?> agmc && agmc.ContainsKey("agents"))
            agentDescriptors.AddRange(A2A.ParseAgentsConfig(agmc.Get("agents")));

        var agentToolLists = await Task.WhenAll(agentDescriptors.Select(async a =>
        {
            try { return await A2A.AgentTools(a).ConfigureAwait(false); }
            catch (Exception e)
            {
                Console.Error.WriteLine($"[toolnexus] A2A agent \"{a.Card}\" failed: {e.Message}");
                return new List<ITool>();
            }
        })).ConfigureAwait(false);
        var agentTools = agentToolLists.SelectMany(x => x).ToList();

        // A2A inbound profile: a top-level `a2a` block on a parsed config object (mirrors
        // builtins/agents). ServeAsync prefers its inline `a2a` over this.
        A2AConfig? a2aConfig = null;
        if (opts.McpConfig is IDictionary<string, object?> a2amc && a2amc.ContainsKey("a2a"))
            a2aConfig = ParseA2AConfig(a2amc.Get("a2a"));

        // MCP inbound profile: a top-level `mcpServer` (singular) block — distinct from the
        // client-side `mcpServers` block. ServeAsync prefers its inline `mcp` over this.
        MCPServeConfig? mcpServerConfig = null;
        if (opts.McpConfig is IDictionary<string, object?> mcpmc && mcpmc.ContainsKey("mcpServer"))
            mcpServerConfig = ParseMcpServeConfig(mcpmc.Get("mcpServer"));

        return new Toolkit(mcp, skill, builtins, agentTools, extras, a2aConfig, mcpServerConfig);
    }

    /// <summary>Coerce a top-level <c>mcpServer</c> config value (an <see cref="MCPServeConfig"/> or a parsed dict) into an <see cref="MCPServeConfig"/>.</summary>
    private static MCPServeConfig? ParseMcpServeConfig(object? value)
    {
        if (value is MCPServeConfig cfg) return cfg;
        if (value is not IDictionary<string, object?> map) return null;
        var result = new MCPServeConfig
        {
            Name = map.Get("name") as string,
            Version = map.Get("version") as string,
        };
        if (map.Get("tools") is IEnumerable<object?> tools)
            result.Tools = tools.Select(t => t?.ToString() ?? "").ToList();
        return result;
    }

    /// <summary>Coerce a top-level <c>a2a</c> config value (an <see cref="A2AConfig"/> or a parsed dict) into an <see cref="A2AConfig"/>.</summary>
    private static A2AConfig? ParseA2AConfig(object? value)
    {
        if (value is A2AConfig cfg) return cfg;
        if (value is not IDictionary<string, object?> map) return null;
        var result = new A2AConfig
        {
            Name = map.Get("name") as string,
            Description = map.Get("description") as string,
            Version = map.Get("version") as string,
            Store = map.Get("store"),
        };
        if (map.Get("skills") is IEnumerable<object?> skills)
            result.Skills = skills.Select(s => s?.ToString() ?? "").ToList();
        if (map.Get("provider") is IDictionary<string, object?> prov)
            result.Provider = new A2AProvider
            {
                Organization = prov.Get("organization")?.ToString() ?? "",
                Url = prov.Get("url")?.ToString() ?? "",
            };
        return result;
    }

    private void Add(ITool t)
    {
        if (_byName.ContainsKey(t.Name))
        {
            Console.Error.WriteLine($"[toolnexus] duplicate tool name \"{t.Name}\" — keeping first");
            return;
        }
        _byName[t.Name] = t;
    }

    public List<ITool> Tools() => _byName.Values.ToList();

    public ITool? Get(string name) => _byName.GetValueOrDefault(name);

    /// <summary>Add native/http/custom tools at runtime. First-name-wins; warn on dup. Chainable.</summary>
    public Toolkit Register(params ITool[] tools)
    {
        foreach (var t in tools) Add(t);
        return this;
    }

    /// <summary>
    /// Fetch a remote A2A agent's card at runtime and register its skills as tools.
    /// First-name-wins dedupe. Chainable.
    /// </summary>
    public async Task<Toolkit> AddAgentAsync(Agent ag)
    {
        var tools = await A2A.AgentTools(ag).ConfigureAwait(false);
        return Register(tools.ToArray());
    }

    /// <summary>Add a remote A2A agent by bare card URL (with optional headers/timeout/pollEvery).</summary>
    public Task<Toolkit> AddAgentAsync(string cardUrl, Agent? opts = null)
    {
        var ag = opts == null
            ? new Agent { Card = cardUrl }
            : new Agent { Card = cardUrl, Headers = opts.Headers, Timeout = opts.Timeout, PollEvery = opts.PollEvery };
        return AddAgentAsync(ag);
    }

    /// <summary>Options for <see cref="ServeAsync"/> — the host loop client plus the optional A2A profile.</summary>
    public sealed class ServeOptions
    {
        /// <summary>The unified LLM client that fulfils inbound tasks (<c>client.Run(text, toolkit)</c>).</summary>
        public LlmClient Client { get; set; } = null!;

        /// <summary>Opt-in A2A profile; falls back to a top-level <c>a2a</c> config block when null.</summary>
        public A2AConfig? A2A { get; set; }

        /// <summary>Fires on each served Task's terminal state with the <see cref="LlmClient.RunResult"/> telemetry.</summary>
        public OnTask? OnTask { get; set; }

        /// <summary>Opt-in MCP inbound profile; mounts a streamable-HTTP MCP server at <c>POST /mcp</c>. Falls back to a top-level <c>mcpServer</c> config block when null.</summary>
        public MCPServeConfig? Mcp { get; set; }

        /// <summary>Fires on each inbound MCP <c>tools/call</c>.</summary>
        public OnCall? OnCall { get; set; }
    }

    /// <summary>
    /// Serve this toolkit as an agent over HTTP. When the <c>a2a</c> profile is present
    /// (inline, or a top-level <c>a2a</c> config block the toolkit was built from), it mounts
    /// the A2A Agent Card (<c>/.well-known/agent-card.json</c>, built from skills) and a
    /// JSON-RPC endpoint (<c>SendMessage</c> + <c>GetTask</c>) fulfilled by the client. A message's
    /// A2A <c>contextId</c> keys the conversation via <c>client.Ask</c>, so a peer's turns are
    /// remembered through the client's <see cref="IConversationStore"/>.
    /// When <c>a2a</c> is absent, no A2A routes are mounted. Returns a stoppable handle.
    /// </summary>
    public Task<ServeHandle> ServeAsync(string addr, ServeOptions opts)
    {
        var a2a = opts.A2A ?? _a2aConfig;
        var mcp = opts.Mcp ?? _mcpServerConfig;
        var skills = _skill != null ? _skill.Skills.Values.ToList() : new List<SkillSource.SkillInfo>();
        return A2AServer.StartAsync(
            addr,
            a2a,
            skills,
            // A message's A2A contextId keys the conversation via client.Ask, so a peer's turns
            // are remembered through the client's IConversationStore; no contextId ⇒ stateless run.
            (text, contextId) => contextId != null
                ? opts.Client.AskAsync(text, this, contextId)
                : opts.Client.RunAsync(text, this),
            opts.OnTask,
            mcp,
            mcp != null ? Tools() : null,
            opts.OnCall);
    }

    public async Task<ToolResult> ExecuteAsync(string name, IDictionary<string, object?>? args, ToolContext? ctx = null)
    {
        if (!_byName.TryGetValue(name, out var tool))
            return ToolResult.Error($"Unknown tool: {name}");
        return await tool.ExecuteAsync(args ?? new Dictionary<string, object?>(), ctx).ConfigureAwait(false);
    }

    /// <summary>Markdown skill catalog for the system prompt.</summary>
    public string SkillsPrompt() => _skill?.Prompt() ?? "";

    /// <summary>server -&gt; "connected" | "failed" | "disabled".</summary>
    public IReadOnlyDictionary<string, string> McpStatus()
        => _mcp?.Status ?? new Dictionary<string, string>();

    public List<Dictionary<string, object?>> ToOpenAI() => Adapters.ToOpenAI(Tools());
    public List<Dictionary<string, object?>> ToAnthropic() => Adapters.ToAnthropic(Tools());
    public List<Dictionary<string, object?>> ToGemini() => Adapters.ToGemini(Tools());

    public async ValueTask DisposeAsync()
    {
        if (_mcp != null) await _mcp.DisposeAsync().ConfigureAwait(false);
    }
}
